kyo-core/shared/src/main/scala/kyo/Hub.scala

```diff
@@ -74,74 +83,108 @@ class Hub[A] private[kyo] (
+    /** Closes the Hub and returns any remaining messages.
+      *
+      * When closed:
+      *   - The Hub stops accepting new messages
+      *   - All listeners are automatically closed
+      *   - Any blocked publishers are unblocked with a Closed failure
+      *   - All subsequent operations will fail with Closed
       *
       * @return
-      *   a Maybe containing any remaining elements in the Hub
+      *   a Maybe containing any messages that were in the Hub's buffer at the time of closing. Returns Absent if the close operation fails
+      *   (e.g., if the Hub was already closed).
       */
     def close(using frame: Frame): Maybe[Seq[A]] < IO =
-        fiber.interruptDiscard(Result.Panic(Closed("Hub", initFrame))).andThen {
+        fiber.interruptDiscard(Result.Failure(Closed("Hub", initFrame))).andThen {
             ch.close.map { r =>
                 IO {
-                    val array = listeners.toArray()
-                    discard(listeners.removeIf(_ => true))
-                    Loop.indexed { idx =>
-                        if idx == array.length then Loop.done
-                        else
-                            array(idx).asInstanceOf[Channel[A]].close
-                                .map(_ => Loop.continue)
-                    }.andThen(r)
+                    val l = Chunk.from(listeners.toArray()).asInstanceOf[Chunk[Listener[A]]]
+                    discard(listeners.clear())
+                    Kyo.foreachDiscard(l)(_.child.close.unit).andThen(r)
                 }
             }
         }
 
-    /** Creates a new listener for this Hub with default buffer size.
+    /** Creates a new listener for this Hub with the default buffer size.
       *
       * @return
-      *   a new Listener
+      *   a new Listener that will receive messages from the Hub
       */
     def listen(using Frame): Listener[A] < (IO & Abort[Closed] & Resource) =
-        listen(0)
+        listen(DefaultBufferSize)
 
-    /** Creates a new listener for this Hub with specified buffer size.
+    /** Creates a new listener for this Hub with the specified buffer size.
       *
       * @param bufferSize
-      *   the size of the buffer for the new listener
+      *   the size of the buffer for the new listener. When full, it may cause hub backpressure affecting all listeners
       * @return
-      *   a new Listener
+      *   a new Listener with the specified buffer size
       */
     def listen(bufferSize: Int)(using frame: Frame): Listener[A] < (IO & Abort[Closed] & Resource) =
+        listen(bufferSize, _ => true)
+
+    /** Creates a new listener for this Hub with a custom filter predicate.
+      *
+      * @param filter
+      *   a predicate function that determines which messages this listener receives
+      * @return
+      *   a new Listener that will only receive messages matching the filter
+      */
+    def listen(filter: A => Boolean)(using frame: Frame): Listener[A] < (IO & Abort[Closed] & Resource) =
```

Comment:
new feature: listeners can provide a function to select messages they're interested in